### ISSUE TITLE
AwesomeWM vs screenshot tools: no window frame captured

### ISSUE THREAD

**Issue by Jajauma** [awesomeWM/awesome#980]
# TL;DR
Hi, I can't capture client window frames using common screenshot creation utilities available.

# Full story
I'm expiriencing bizzare issues while trying to make screenshots of active windows (clients?) under AwesomeWM. There is quite a stock of different screenshot creation utilites available out there --- `import`, `scrot`, mate-screenshot/gnome-screenshot, GIMP --- just to name a few. 

The ill feature they have in common while running under AwesomeWM is lack of the ability to reasonably capture window decorations (dark 1px stroke around a window and a titlebar on top of the client area) for some reason. Here is a quick low-down on my tests under latest AwesomeWM git using default configuration.

# The wierd behaviour summary
## mate-screenshot/gnome-screenshot (`mate-screenshot -i`)
![matescreenshot-ui](https://cloud.githubusercontent.com/assets/11892063/16409259/a7be5574-3d2d-11e6-9087-d1ab5e8eb85b.png)
![gnomescreenshot-ui](https://cloud.githubusercontent.com/assets/11892063/16410894/a03797b8-3d35-11e6-87c5-483c8581d9d4.png)

Both crash with SIGSEGV when I put the "Include the window border" tick, press "Take Screenshot" and select the window, and don't produce any screenshot therefore. Both work flawlessly under ratpoison and openbox.

## GIMP (File → Create → Screenshot)
![gimp-ui](https://cloud.githubusercontent.com/assets/11892063/16409359/27f9d5ba-3d2e-11e6-9fef-9affa244ce7f.png)
Screenshot taken by GIMP has a titlebar, but lacks the window 1px frame, see the example.
![gimp-test](https://cloud.githubusercontent.com/assets/11892063/16409432/9083705a-3d2e-11e6-95c8-2bed45d2ac9f.png)

## `import` (`import -border test.png`)
Screenshot taken by `import` lacks both titlebar and window 1px frame, see the example.
![import-test](https://cloud.githubusercontent.com/assets/11892063/16409509/e8867338-3d2e-11e6-85bc-392a3ac3f33d.png)

## `scrot` (`scrot -b -s test.png`)
Screenshot taken by `scrot` has a titlebar, but lacks the window 1px frame (just like the GIMP one does), see the example.
![scrot-test](https://cloud.githubusercontent.com/assets/11892063/16409557/5026a4d6-3d2f-11e6-8195-b976e79aba8f.png)

# Conclusion
The real question appears --- what part of my system is broken? Well, apparently it's the screenshoter tools which fail in the first place, but the grim ratio of four to one makes me submit not theirs but AwesomeWM's issue first.

**Comment by psychon** [comment]
I love the segfaults. I guess you should report that upstream. :-)
If you want to work around them on the awesome side, then launch awesome with `--no-argb` and I guess that things will work. For the upstream bug report: Awesome creates its frame window with depth 32 if possible while most "normal" windows have a depth of 24. This causes lots of X11-complications. You should experiment a bit with `urxvt -depth 32` and see if this makes some problems disappear (and could cause similar problems under other WMs).

For the window border: I don't think we can do anything about this. X11 provides "real" window borders (which only have a single color). Apparently awesome and e.g. dwm use them, but everything else does not use them; likely because of the "only a single color" limitation. There is nothing I could do about that (besides stopping to use them, but no thanks).

**Comment by Jajauma** [comment]
Thanks for the prompt reply, sadly neither `--no-argb` nor `urxvt -depth 32` did any difference for me. Mate-screenshot still crashes at some stiff line like `*dest_pixels++ = *src_pixels++;`, I hope this could be resolved upstream.

It's such a shame there is no way to capture that _real_ X11 border. Pictures look just cropped without a stroke around. If I understand you correctly, the real X11 border is not considered as part of the window, so the "usual" contents grabbing tools won't nail it anyway. Maybe running one of those "compositing" thingies used to deliver fancy graphics (transparency and such) could be of any help?

**Comment by Jajauma** [comment]
I guess I better close it since there is no really a bug from the AwesomeWM side regarding that 1px border.

**Comment by blueyed** [comment]
You could try if Compton makes a difference, yes.
Have you reported the segfaults upstream? Please provide a link here then.

**Comment by Jajauma** [comment]
Somebody already came across the similar issue https://github.com/mate-desktop/mate-utils/issues/147 and I just contributed with a comment there, but nobody really seems to be bothered much, and I'm not nearly around being able to debug that crash all by myself. 

Compton didn't make any difference as well, so I went all faint-heartedly and just gave up to capture that ghost AwesomeWM 1px border in the end.

**Comment by psychon** [comment]
I couldn't reproduce the crash for mate-screenshot here with `mate-screenshot -b -w`. Valgrind complains about lots of leaks, but not about invalid memory access.

**Comment by Jajauma** [comment]
It's a bummer, valgrind carps and croaks while running mate-screenshot under AwesomeWM. Well I guess it's just only poor me again so there is no big deal about that (I still can capture the whole screen and crop the picture after all).
```
$ valgrind /usr/local/bin/mate-screenshot -w -b
==4426== Memcheck, a memory error detector
==4426== Copyright (C) 2002-2013, and GNU GPL'd, by Julian Seward et al.
==4426== Using Valgrind-3.10.0 and LibVEX; rerun with -h for copyright info
==4426== Command: /usr/local/bin/mate-screenshot -w -b
==4426== 
==4426== 
==4426== Process terminating with default action of signal 11 (SIGSEGV)
==4426==  Bad permissions for mapped region at address 0x1BD15352
==4426==    at 0x40BB34: screenshot_get_pixbuf (screenshot-utils.c:956)
==4426==    by 0x409846: finish_prepare_screenshot (mate-screenshot.c:898)
==4426==    by 0x409846: check_file_done (mate-screenshot.c:955)
==4426==    by 0x6567A2E: mainloop_proxy_func (gioscheduler.c:200)
==4426==    by 0x790D7A9: g_main_dispatch (gmain.c:3109)
==4426==    by 0x790D7A9: g_main_context_dispatch (gmain.c:3708)
==4426==    by 0x790DAF7: g_main_context_iterate.isra.24 (gmain.c:3779)
==4426==    by 0x790DDC9: g_main_loop_run (gmain.c:3973)
==4426==    by 0x5AE2996: gtk_main (gtkmain.c:1268)
==4426==    by 0x40804E: main (mate-screenshot.c:1456)
==4426== 
==4426== HEAP SUMMARY:
==4426==     in use at exit: 14,758,740 bytes in 5,267 blocks
==4426==   total heap usage: 20,505 allocs, 15,238 frees, 378,308,391 bytes allocated
==4426== 
==4426== LEAK SUMMARY:
==4426==    definitely lost: 192 bytes in 4 blocks
==4426==    indirectly lost: 5,983,668 bytes in 1 blocks
==4426==      possibly lost: 8,271,847 bytes in 289 blocks
==4426==    still reachable: 454,969 bytes in 4,700 blocks
==4426==         suppressed: 0 bytes in 0 blocks
==4426== Rerun with --leak-check=full to see details of leaked memory
==4426== 
==4426== For counts of detected and suppressed errors, rerun with: -v
==4426== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 1 from 1)
Убито
```
while both doing well under openbox and take the darn laboured screenshot.
